File: usage_example/cpp/step1/src/main.cpp
#include <stdio.h>
#include "c3ga.h"

using namespace c3ga;

/*
This small example creates a line and a plane from points.
It then computes the intersection point of the line and the plane.

In this step, all GA variables are stored in 'mv' (the multivector type).
*/
int main(int argc, char *argv[]) {

	// get five points
	mv linePt1 = cgaPoint(1.0f, 0.0f, 0.0f);
	mv linePt2 = cgaPoint(1.0f, 1.0f, 0.0f);
	mv planePt1 = cgaPoint(1.0f, 2.0f, 0.0f);
	mv planePt2 = cgaPoint(1.0f, 2.0f, 1.0f);
	mv planePt3 = cgaPoint(0.0f, 2.0f, 1.0f);

	// output text the can be copy-pasted into GAViewer
	// warning: c_str() is not multi-threading safe (for that, use toString() instead).
	printf("linePt1 = %s,\n", linePt1.c_str()); 
	printf("linePt2 = %s,\n", linePt2.c_str());
	printf("planePt1 = %s,\n", planePt1.c_str());
	printf("planePt2 = %s,\n", planePt2.c_str());
	printf("planePt3 = %s,\n", planePt3.c_str());

	// create line and plane out of points
	mv L = linePt1 ^ linePt2 ^ ni;
	mv P = planePt1 ^ planePt2 ^ planePt3 ^ ni;

	// output text the can be copy-pasted into GAViewer
	printf("L = %s,\n", L.c_str());
	printf("P = %s,\n", P.c_str());

	// compute intersection of line and plane
	mv intersection = dual(L) << P;

	printf("intersection = %s,\n", intersection.c_str());

	return 0;
}

File: usage_example/java/Step1/src/Main.java
import C3ga_pkg.C3ga;
import C3ga_pkg.Mv;


public class Main {

    /**
    This small example creates a line and a plane from points.
    It then computes the intersection point of the line and the plane.

    In this step, all GA variables are stored in 'mv' (the multivector type).
    */
	public static void main(String[] args) {
		
        // get five points
        Mv linePt1 = new Mv(C3ga.cgaPoint(1.0f, 0.0f, 0.0f));
        Mv linePt2 = new Mv(C3ga.cgaPoint(1.0f, 1.0f, 0.0f));
        Mv planePt1 = new Mv(C3ga.cgaPoint(1.0f, 2.0f, 0.0f));
        Mv planePt2 = new Mv(C3ga.cgaPoint(1.0f, 2.0f, 1.0f));
        Mv planePt3 = new Mv(C3ga.cgaPoint(0.0f, 2.0f, 1.0f));

        // output text the can be copy-pasted into GAViewer
        System.out.println("linePt1 = " + linePt1 + ",");
        System.out.println("linePt2 = " + linePt2 + ",");
        System.out.println("planePt1 = " + planePt1 + ",");
        System.out.println("planePt2 = " + planePt2 + ",");
        System.out.println("planePt3 = " + planePt3 + ",");

        // create line and plane out of points
        Mv L = C3ga.op(linePt1, C3ga.op(linePt2, C3ga.ni));
        Mv P = C3ga.op(planePt1, C3ga.op(planePt2, C3ga.op(planePt3, C3ga.ni)));

        // output text the can be copy-pasted into GAViewer
        System.out.println("L = " + L + ",");
        System.out.println("P = " + P + ",");

        // compute intersection of line and plane
        Mv intersection = C3ga.lc(C3ga.dual(L), P);

        System.out.println("intersection = " + intersection + ",");
	}
}
